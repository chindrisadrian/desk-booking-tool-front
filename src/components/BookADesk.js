import React from 'react';
import { setTab, setFloor, setLimit, setBookingInfo } from '../actions';
import { connect } from 'react-redux';
import '../styles/base.css';
import { DateInput } from 'semantic-ui-calendar-react';
import { Form } from 'semantic-ui-react';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import uuid from 'uuid';
import moment from 'moment';
import BookingRestrictionService from '../services/bookingRestrictionService';
import FloorService from '../services/floorService';
import history from '../history';

class Book extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      date: '',
      office: '',
      floor: '',
      err: false,
      floors: [],
    };

    this.service = new FloorService();
    this.resService = new BookingRestrictionService();
    this.getFloors();
    this.getLimits();
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setTab('book'));
  }

  getFloors = async () => {
    const floors = await this.service.getFloors({});
    this.setState({ floors });
  };

  getLimits = async () => {
    const { dispatch } = this.props;
    const limit = await this.resService.getRestrictions();
    dispatch(setLimit(limit[0]));
  };

  handleDateChange = (e, { value }) => {
    this.setState({ date: `${value}` });
  };

  handleOffice = (e) => {
    this.setState({ office: e.target.value });
  };

  handleFloor = (e) => {
    this.setState({ floor: e.target.value });
  };

  maxDate = () => {
    let { limit } = this.props;
    if (!limit) limit = { continuous: 0, ahead: 0 };
    const maxDate = moment();
    maxDate.add(limit.ahead, 'days');
    return maxDate;
  };

  handleCancel = () => {
    this.setState({
      date: '',
      office: '',
      floor: '',
    });
  };

  onNext = async () => {
    const { date, office, floor, floors } = this.state;
    const { dispatch } = this.props;
    await this.setState({ err: false });
    if (date === '' || office === '' || floor === '') {
      await this.setState({ err: true });
    } else {
      const { image } = floors.find(
        (elm) => elm.office == office && elm.floor == floor
      );
      const obj = {
        office,
        floor,
        date,
        image,
      };
      const selectedFloor = floors.find(
        (f) => f.office === office && f.floor === floor
      );
      dispatch(setFloor(selectedFloor));
      dispatch(setBookingInfo(obj));
      history.push('/setBooking');
    }
  };

  render() {
    const { err, floors } = this.state;
    let { limit } = this.props;
    this.maxDate();
    if (!limit) limit = { continuous: 0, ahead: 0 };
    const offices = [];
    for (const i in floors) {
      if (!offices.includes(floors[i].office)) offices.push(floors[i].office);
    }
    return (
      <div>
        <div className="details"> Details </div>
        <div className="note">
          You can book as much as {limit.continuous} continuous days and a
          maximum of {limit.ahead} days ahead
        </div>
        <Form className="detail-from">
          <FormControl variant="outlined" className="my-input">
            <InputLabel
              id="demo-simple-select-outlined-label"
              className="my-text"
            >
              Office
            </InputLabel>
            <Select
              labelId="demo-simple-select-outlined-label"
              className="my-text"
              id="demo-simple-select-outlined"
              onChange={this.handleOffice}
              label="Office"
              value={this.state.office}
            >
              {offices.map((office) => (
                <MenuItem value={office} className="my-text" key={uuid()}>
                  {office}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
          <FormControl variant="outlined" className="my-input">
            <InputLabel
              id="demo-simple-select-outlined-label"
              className="my-text"
            >
              Floor
            </InputLabel>
            <Select
              labelId="demo-simple-select-outlined-label"
              className="my-text"
              id="demo-simple-select-outlined"
              onChange={this.handleFloor}
              label="Floor"
              value={this.state.floor}
            >
              {this.state.office ? (
                floors.map((floor) =>
                  floor.office === this.state.office ? (
                    <MenuItem
                      value={floor.floor}
                      className="my-text"
                      key={uuid()}
                    >
                      {floor.floor}
                    </MenuItem>
                  ) : (
                    ''
                  )
                )
              ) : (
                <MenuItem value="" className="my-text" disabled>
                  No office selected
                </MenuItem>
              )}
            </Select>
          </FormControl>
          <DateInput
            name="datesRange"
            className="my-input my-text"
            placeholder="Select a Date"
            value={this.state.date}
            iconPosition="left"
            onChange={this.handleDateChange}
            maxDate={this.maxDate()}
          />
          <div className="ui button cancel" onClick={this.handleCancel}>
            Reset
          </div>

          <div className="ui blue button" onClick={this.onNext}>
            Next
          </div>
        </Form>
        {err ? (
          <div className="err">
            ! Please fill in all fields before continuing
          </div>
        ) : (
          ''
        )}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  limit: state.limit,
});

export default connect(mapStateToProps)(Book);
