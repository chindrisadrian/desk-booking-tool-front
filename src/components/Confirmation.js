import React from 'react';
import '../styles/Restrictions.css';
import { connect } from 'react-redux';
import { Dialog } from '@material-ui/core';
import { Card, Header } from 'semantic-ui-react';
import history from '../history';
import { setTab } from '../actions';
import BookingService from '../services/bookingService';

class Confirmation extends React.Component {
  constructor(props) {
    super(props);

    this.state = { done: false, cont: false, same: false };
    this.service = new BookingService();
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setTab('book'));
    this.checkContinuous();
  }

  componentDidUpdate() {}

  sortByDate = (d1, d2) => {
    if (parseInt(d1.date.substr(0, 4)) < parseInt(d2.date.substr(0, 4))) {
      return -1;
    }
    if (parseInt(d1.date.substr(0, 4)) > parseInt(d2.date.substr(0, 4))) {
      return 1;
    }
    if (parseInt(d1.date.substr(5, 2)) < parseInt(d2.date.substr(5, 2))) {
      return -1;
    }
    if (parseInt(d1.date.substr(5, 2)) > parseInt(d2.date.substr(5, 2))) {
      return 1;
    }
    if (parseInt(d1.date.substr(8, 2)) < parseInt(d2.date.substr(8, 2))) {
      return -1;
    }
    if (parseInt(d1.date.substr(8, 2)) > parseInt(d2.date.substr(8, 2))) {
      return 1;
    }
    return 0;
  };

  checkContinuous = async () => {
    const { email, booking, limit } = this.props;
    let bookings = await this.service.getBookings({
      email: email,
    });
    const bookDate = `${booking.date
      .split('-')
      .reverse()
      .join('-')}T00:00:00.000Z`;
    if (bookings.find((e) => e.date === bookDate) !== undefined) {
      this.setState({ same: true });
      return;
    }

    bookings.push({
      date: bookDate,
      email: identification.email,
    });
    bookings = bookings.sort(this.sortByDate);
    let curIndex = bookings.findIndex((e) => e.date === bookDate);
    let forward = 0;
    let backward = 0;
    while (1) {
      if (curIndex === bookings.length - 1) {
        break;
      }
      const dateA = new Date(bookings[curIndex].date);
      dateA.setDate(dateA.getDate() + 1);
      const A = dateA.toString().substr(0, 16);
      const dateB = new Date(bookings[curIndex + 1].date);
      const B = dateB.toString().substr(0, 16);
      if (A == B) {
        console.log('yes');
        forward += 1;
        curIndex += 1;
      } else {
        break;
      }
    }
    curIndex = bookings.findIndex((e) => e.date === bookDate);
    while (1) {
      if (curIndex === 0) {
        break;
      }
      const dateA = new Date(bookings[curIndex].date);
      dateA.setDate(dateA.getDate() - 1);
      const A = dateA.toString().substr(0, 16);
      const dateB = new Date(bookings[curIndex - 1].date);
      const B = dateB.toString().substr(0, 16);
      if (A === B) {
        backward += 1;
        curIndex -= 1;
      } else {
        break;
      }
    }
    const totalContinuous = 1 + forward + backward;
    if (totalContinuous > limit.continuous) {
      this.setState({ cont: true });
    }
  };

  onUserSubmit = async (event) => {
    const { desk, booking, email } = this.props;
    await this.service.postBooking({
      email: email,
      office: booking.office,
      date: booking.date.split('-').reverse().join('-'),
      floor: booking.floor,
      desk,
    });
    this.setState({ done: true });
  };

  done = () => {
    location.reload();
    return true;
  };

  goBack = () => {
    history.push('/setBooking');
  };

  render() {
    const { booking, desk, limit, email } = this.props;
    const { same, cont } = this.state;

    if (desk === null) {
      history.push('/setBooking');
      return null;
    }

    return (
      <div>
        <div className="details">Booking Confirmation</div>
        <div className="conf-data">
          <div className="conf">
            <b>Email:</b> {email}{' '}
          </div>

          <div className="conf">
            <b>Office:</b> {booking.office}
          </div>
          <div className="conf">
            <b>Floor: </b>
            {booking.floor}
          </div>
          <div className="conf">
            <b>Desk Selection:</b> {desk}
          </div>
          <div className="conf">
            <b>Date:</b> {booking.date}
          </div>
        </div>
        {same ? (
          <div className="warn">
            <p className="warn-text">
              ! You have already made a booking on this day.
            </p>
          </div>
        ) : (
          ''
        )}
        {cont ? (
          <div className="warn">
            <p className="warn-text">
              ! Cannot book more than {limit.continuous} days in a row.
            </p>
          </div>
        ) : (
          ''
        )}

        <div>
          <div className="ui button " onClick={this.goBack}>
            Back
          </div>
          <div
            className={`ui blue button ${same || cont ? 'disabled' : ''}`}
            onClick={this.onUserSubmit}
          >
            Confirm
          </div>
        </div>
        <Dialog
          open={this.state.done}
          aria-labelledby="form-dialog-title"
          style={{ width: '100%', maxWidth: 'xl' }}
        >
          <Card className="my-card3">
            <Header className="conf-header">
              Thanks for using the Desk Booking Tool.
            </Header>
            <div className="thanks">
              Your booking for desk {desk} on floor {booking.floor} at{' '}
              {booking.office} on {booking.date} is saved{' '}
            </div>
            <div className="thanks">
              A confirmation email has been sent to {email}
            </div>
            <div className="ui blue button" id="thanks-btn" onClick={this.done}>
              Done
            </div>
          </Card>
        </Dialog>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  identification: state.identification,
  booking: state.booking,
  desk: state.desk,
  limit: state.limit,
  email: state.user,
});

export default connect(mapStateToProps)(Confirmation);
