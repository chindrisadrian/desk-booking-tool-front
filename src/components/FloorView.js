import React from 'react';
import uuid from 'uuid';
import FloorService from '../services/floorService';
import BookingService from '../services/bookingService';
import Image from './image.js';
import Legend from './legend';
import Logo from './logo';

export default class FloorView extends React.Component {
  constructor(props) {
    super(props);

    this.state = { floor: null, booked: {}, selected: {} };

    this.floorService = new FloorService();
    this.bookingService = new BookingService();

    this.getBookings();
  }

  getFloor = async () => {
    const { floor, office } = this.props;
    const theFloor = await this.floorService.getFloors({
      floor,
      office,
    });
    this.setState({ floor: theFloor[0] });
  };

  getBookings = async () => {
    await this.getFloor();
    const { date, newFloor, office } = this.props;
    const { floor } = this.state;
    const bookings = await this.bookingService.getBookings({
      date,
      office,
      floor: newFloor,
    });
    var booked = {};
    var selected = {};
    for (let i in floor.desks) {
      selected[floor.desks[i].name] = false;
      if (
        bookings.find((booking) => booking.desk === floor.desks[i].name) !==
        undefined
      ) {
        booked[floor.desks[i].name] = true;
      } else {
        booked[floor.desks[i].name] = false;
      }
    }
    selected[this.props.desk] = true;
    this.setState({ booked: booked, selected: selected });
  };

  calcX = (x) => {
    const { floor } = this.state;
    const contX = $(window).width() * 0.6;
    const ratio = contX / floor.width;
    return x * ratio + (25 * contX) / 2000;
  };

  calcY = (x) => {
    const { floor } = this.state;
    const contY = $(window).width() * 0.6 * (floor.height / floor.width);
    const ratio = contY / floor.height;
    return x * ratio + (25 * contY) / 1500;
  };

  render() {
    const { date, office } = this.props;
    const { booked, floor, selected } = this.state;
    if (floor === null) {
      return <Logo />;
    }

    return (
      <div>
        <div className="view-sub-header">
          {date.substr(0, 10)} on {this.props.floor} at {office}
        </div>
        <Legend />
        <div className="pics2" id="container">
          <Image data={floor.image.data.data} />
          {floor.desks.map((desk) => (
            <div
              className={`desk-icon ${
                !selected[desk.name]
                  ? booked[desk.name]
                    ? 'booked'
                    : desk.bookable
                    ? 'avail'
                    : 'unbookable'
                  : 'selected'
              }`}
              key={uuid()}
              style={{ left: this.calcX(desk.x), top: this.calcY(desk.y) }}
            >
              <div className="desk-icon-div">
                {desk.name} <br />
                <i className="laptop icon desk-i" />{' '}
              </div>
            </div>
          ))}
        </div>
      </div>
    );
  }
}
