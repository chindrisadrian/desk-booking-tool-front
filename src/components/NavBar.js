import React, { Component } from 'react';
import '../styles/Navbar.css';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class Navbar extends React.Component {
  constructor(props) {
    super(props);
    this.active = 'item active';
    this.inactive = 'item';
  }

  render() {
    const { tab } = this.props;

    return (
      <div>
        <Link to="/feedback" className="feedback-link">
          <div className="site-feedback">
            {' '}
            <i class="wrench icon" />
            Site Feedback{' '}
          </div>
        </Link>
        <div className="site-header">Desk Booking Tool</div>
        <div className="site-subheader">User Panel</div>

        <div className="custom-divider"> </div>
        <div className="ui blue six item menu">
          <Link
            to="/BookADesk"
            className={tab === 'book' ? this.active : this.inactive}
          >
            Book A Desk
          </Link>
          <Link
            className={tab === 'view' ? this.active : this.inactive}
            to="/ViewPreviousBookings"
          >
            View My Previous Bookings
          </Link>
          <Link
            to="/ViewAllBookings"
            className={tab === 'viewAll' ? this.active : this.inactive}
          >
            View All Bookings
          </Link>
          <Link
            className={tab === 'res' ? this.active : this.inactive}
            to="/userRestrictions"
          >
            Timing Restrictions
          </Link>
          <Link
            className={tab === 'desks' ? this.active : this.inactive}
            to="/restrict"
          >
            Restrict Desks
          </Link>
          <Link
            className={tab === 'add' ? this.active : this.inactive}
            to="/addFloor"
          >
            Add a Floor
          </Link>
        </div>
        <div className="custom-divider"> </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  tab: state.tab,
});

export default connect(mapStateToProps)(Navbar);
