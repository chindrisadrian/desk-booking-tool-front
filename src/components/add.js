import React from 'react';
import { connect } from 'react-redux';
import './../styles/add.css';
import { setTab, setImage } from '../actions';
import { Button } from '@material-ui/core';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

/** Upload floorplan pdf page */
class Add extends React.Component {
  /** */
  constructor(props) {
    super(props);
    /** No image has been uploaded yes */
    this.state = { image: null };
  }
  /**  Sets the tab for the navbar */
  componentDidMount() {
    this.props.dispatch(setTab('add'));
  }

  /** Create a URL for the image to display on the next page, and write the image data to redux */
  readFile = (e) => {
    var imageURL = URL.createObjectURL(e.target.files[0]);
    this.setState({ image: imageURL });
    this.props.dispatch(
      setImage({ url: imageURL, image: Array.from(e.target.files)[0] })
    );
  };
  /** */
  render() {
    return (
      <div>
        <div className="details">Upload Your Floor Plan</div>
        <div className="choose">Choose a .jpeg or .png </div>
        <input
          type="file"
          onChange={this.readFile}
          className="inpt"
          accept=".jpg,.png,.jpeg,.JPG"
        />

        {this.state.image ? (
          <div>
            <img src={this.state.image} alt="floorplan" className="preview" />
            <Link to="/floorBuilder">
              <Button
                variant="contained"
                color="primary"
                onClick={this.makeDesk}
                className="upload-btn mui-btn"
                onClick={this.next}
              >
                Next
              </Button>
            </Link>
          </div>
        ) : (
          ''
        )}
      </div>
    );
  }
}

Add.propTypes = {
  dispatch: PropTypes.any,
};

export default connect()(Add);
