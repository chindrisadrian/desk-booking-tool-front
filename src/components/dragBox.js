import React, { Component } from 'react';
import { connect } from 'react-redux';
import Draggable from 'react-draggable';
import FloorService from '../services/floorService';
import '../styles/base.css';
import uuid from 'uuid';
import { Redirect } from 'react-router-dom';
import { Button } from '@material-ui/core';
import { setTab } from '../actions';
import { Dialog, Select, MenuItem, InputLabel } from '@material-ui/core';
import { Card, Header } from 'semantic-ui-react';
import FormData from 'form-data';
import axios from 'axios';
import PropTypes from 'prop-types';

/** Gets the URL to make axios call from this file */
const URL = process.env.REACT_APP_API;
/** The component to drag desks and make a floor  */
class DragBox extends Component {
  /** */
  constructor(props) {
    super(props);
    /** Initialize state*/
    this.state = {
      disabled: false,
      name: '',
      floors: [],
      desks: [],
      autofill: false,
      openAuto: false,
      confirm: false,
      finalDesks: [],
      floorName: '',
      office: '',
    };
    /** To get the floors for when the user saves */
    this.service = new FloorService();
    this.getFloors();
    /** Style for icons that are bokable desks */
    this.bookStyle = {
      width: 90,
      border: '1px solid grey',
      padding: '2px 6px 1px 6px',
      backgroundColor: 'rgba(180,180,180,0.9)',
      borderRadius: '6px',
    };
    /** Style for icons that are unbookable */
    this.unbookStyle = {
      width: 90,
      border: '1px solid lightgrey',
      padding: '2px 6px 1px 6px',
      backgroundColor: 'rgba(150,255,150,0.95)',
      borderRadius: '6px',
    };
    /** Moves the deleted icons off the screen */
    this.deletedStyle = {
      width: 90,
      opacity: 0,
      zIndex: -10,
    };
  }
  /** Set the tab for the navbar */
  componentDidMount() {
    this.props.dispatch(setTab('add'));
  }
  /** Return a new desk */
  makeDesk = () => {
    var desk = {
      disabled: false,
      id: uuid(),
      name: '',
      bookable: true,
      deleted: false,
      x: 0,
      y: 0,
    };
    this.setState({ desks: [...this.state.desks, desk] });
  };
  /** Get floors from the DB  */
  getFloors = async () => {
    const floors = await this.service.getFloors({});
    this.setState({ floors: floors });
  };
  /** Toggle an element between locked and draggable */
  toggleDraggable = (id) => {
    var desks = this.state.desks;
    for (let i in desks) {
      if (desks[i].id === id) {
        desks[i].disabled = !desks[i].disabled;
      }
    }
    this.setState({ desks: desks });
  };
  /** Toggle bookable <---> unbookable  */
  click = (id) => {
    var desks = this.state.desks;
    for (let i in desks) {
      if (desks[i].id === id) {
        desks[i].bookable = !desks[i].bookable;
      }
    }
    this.setState({ desks: desks });
  };
  /** Sets a desk to deleted so it will no longer appear */
  delete = (id) => {
    var desks = this.state.desks;
    for (let i in desks) {
      if (desks[i].id === id) {
        desks[i].deleted = true;
      }
    }
    this.setState({ desks: desks });
  };

  /** Value setter */
  rename = (e) => {
    const { id, value } = e.target;
    var desks = this.state.desks;
    for (let i in desks) {
      if (desks[i].id === id) {
        desks[i].name = value;
      }
    }
    this.setState({ desks: desks });
  };

  /** Writes all the info to database using formdata */
  /** Saves the width and height of the floorplan as it was made so that the positions can
   * be scaled for smaller or larger screens on other tabs
   * Uses jQuery to get the position of all the desks, and saves that along with the desk name
   * and bookability
   */
  done = () => {
    var finalDesks = [];
    let cont = document.getElementById('dragContainer').getBoundingClientRect();
    var { desks, autofill } = this.state;
    var x = true;
    var i = 1;
    desks.map((desk) => {
      let newDesk = { bookable: desk.bookable, name: desk.name };
      if (newDesk.name === '' && !desk.deleted) {
        /** Starting with 1, try and set desk names */
        if (autofill) {
          while (1) {
            if (desks.find((d) => d.name == i.toString()) === undefined) {
              newDesk.name = i.toString();
              i = i + 1;
              break;
            }
          }
        } else {
          this.setState({ openAuto: true });
          x = false;
        }
      }
      let elem = document.getElementById(desk.id).getBoundingClientRect();
      newDesk['x'] = elem.x - cont.x - 0.6;
      newDesk['y'] = elem.y - cont.y - 0.6;
      if (!desk.deleted) {
        finalDesks.push(newDesk);
      }
    });
    if (x) {
      this.setState({ confirm: true, finalDesks: finalDesks, autofill: false });
    }
  };

  /** Sets the autofill value to true, so the submit function will just generate desk names */
  autofill = async () => {
    await this.setState({ autofill: true, openAuto: false });
    this.done();
  };

  /** Once everything is verified, make an axios call */
  publish = async () => {
    const { finalDesks, office, floorName } = this.state;
    var containerHeight = document
      .getElementById('underlay')
      .getBoundingClientRect().height;
    var containerWidth = document
      .getElementById('dragContainer')
      .getBoundingClientRect().width;
    let data = new FormData();
    data.append('image', this.props.image.image, 'image');
    data.append('office', office);
    data.append('floor', floorName);
    data.append('width', containerWidth);
    data.append('desks', JSON.stringify(finalDesks));
    data.append('height', containerHeight);
    await axios
      .post(`http://${URL}/floorsFrontend`, data, {
        headers: {
          accept: 'application/json',
          'Accept-Language': 'en-US,en;q=0.8',
          'Content-Type': `multipart/form-data; boundary=${data._boundary}`,
        },
      })
      .then((response) => {
        console.log(response);
      })
      .catch((error) => {
        console.log(error);
      });

    location.reload();
  };

  /** Change the office name on the publish window */
  handleOffice = (e) => {
    this.setState({ office: e.target.value });
  };
  /** */
  render = () => {
    const { desks, floors, confirm, openAuto, office, floorName } = this.state;
    const { image } = this.props;

    if (image === null) {
      return <Redirect to="/addFloor" />;
    }
    const { url } = image;
    var offices = [];

    for (let i in floors) {
      if (!offices.includes(floors[i].office)) offices.push(floors[i].office);
    }

    return (
      <div>
        <div className="details"> Dynamic Floor Plan </div>
        <div className="detail-sub">
          Add desks using the button at the bottom of the page: drag and drop
          them onto the floorplan and enter a label.{' '}
        </div>
        <div id="dragContainer">
          <div className="underlay" id="underlay">
            <img src={url} alt="floorplan" className="floor-pic" />
          </div>

          <Button
            variant="contained"
            color="primary"
            className="mui-btn"
            onClick={this.done}
          >
            Done
          </Button>
          <Button
            variant="contained"
            color="primary"
            className="mui-btn"
            onClick={this.makeDesk}
          >
            <i className="plus icon pls"></i> New Desk
          </Button>
          {desks.map((desk) => (
            <Draggable
              disabled={desk.disabled}
              bounds="parent"
              key={desk.id}
              position={null}
            >
              <div
                style={
                  desk.deleted
                    ? this.deletedStyle
                    : !desk.bookable
                    ? this.bookStyle
                    : this.unbookStyle
                }
                className={
                  desk.disabled ? 'draggable' : 'draggable draggable-active'
                }
                id={desk.id}
              >
                <div className="edit-icon">
                  <div
                    className="my-i left"
                    onClick={() => this.toggleDraggable(desk.id)}
                  >
                    {desk.disabled ? (
                      <i className="hand rock icon"></i>
                    ) : (
                      <i className="lock icon"></i>
                    )}
                  </div>
                  <div
                    className="my-i right"
                    onClick={() => this.delete(desk.id)}
                  >
                    <i className="delete icon"></i>
                  </div>
                </div>
                <div className="icon-div">
                  <i className="laptop icon dbt-icon"></i>
                </div>
                <div className="ui input desk-input">
                  <input
                    type="text"
                    placeholder="Name"
                    onChange={this.rename}
                    value={desk.name}
                  ></input>
                </div>

                <div className="ui checkbox">
                  <input
                    type="checkbox"
                    name="example"
                    onChange={() => this.click(desk.id)}
                    checked={desk.bookable}
                  />
                  <label>Bookable?</label>
                </div>
                <br />
              </div>
            </Draggable>
          ))}
        </div>
        {/** Render the compenent to look at dialog flow. Will open the autfill? window first
         * if not all desks have names, and then the publish dialog
         */}
        {openAuto ? (
          <Dialog
            open={true}
            aria-labelledby="form-dialog-title"
            style={{ width: '100%', maxWidth: 'xl' }}
          >
            <Card className="my-card2">
              <Header className="conf-header">
                Some of Your Desks Don{"'"}t Have Names.
              </Header>
              <div className="thanks">Autofill them numerically?</div>
              <div
                className="ui button"
                id="pub-btn"
                onClick={() => {
                  this.setState({ openAuto: false });
                }}
              >
                Back
              </div>
              <div
                className="ui blue button"
                id="auto-btn"
                onClick={this.autofill}
              >
                Autofill
              </div>
            </Card>
          </Dialog>
        ) : (
          ''
        )}
        {confirm ? (
          <Dialog
            open={true}
            aria-labelledby="form-dialog-title"
            style={{ width: '100%', maxWidth: 'xl' }}
          >
            <Card className="my-card">
              <Header className="conf-header">Publish this floor?</Header>
              <div
                className="ui blue button"
                id="auto-btn"
                onClick={() => {
                  this.setState({ confirm: false });
                }}
              >
                Back
              </div>
              <InputLabel
                id="demo-simple-select-outlined-label"
                className="input-label"
              >
                Select Office
              </InputLabel>
              <Select
                labelId="demo-simple-select-outlined-label"
                className="office-selector"
                id="demo-simple-select-outlined"
                onChange={this.handleOffice}
                label="Office"
                value={this.state.office}
              >
                {offices.map((office) => (
                  <MenuItem value={office} className="my-text" key={uuid()}>
                    {office}
                  </MenuItem>
                ))}
              </Select>
              <div style={{ margin: 'auto' }}> {'//'} </div>
              <div className="ui input id-input office-input">
                <input
                  type="text"
                  placeholder="Enter a New Office Name"
                  onChange={(event) => {
                    this.setState({ office: event.target.value });
                  }}
                  value={this.state.office}
                />
              </div>
              <div className="ui input id-input">
                <input
                  type="text"
                  placeholder="Floor Name"
                  onChange={(event) => {
                    this.setState({ floorName: event.target.value });
                  }}
                  value={this.state.floorName}
                />
              </div>

              <div
                className={
                  office === '' || floorName === ''
                    ? 'ui blue button disabled'
                    : 'ui blue button'
                }
                id="pub-btn"
                onClick={this.publish}
              >
                Confirm
              </div>
            </Card>
          </Dialog>
        ) : (
          ''
        )}
      </div>
    );
  };
}

DragBox.propTypes = {
  image: PropTypes.objectOf(PropTypes.object),
  dispatch: PropTypes.func,
};

/** Redux */
const mapStateToProps = (state) => ({
  image: state.image,
});

export default connect(mapStateToProps)(DragBox);
