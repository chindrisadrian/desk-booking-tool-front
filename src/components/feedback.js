import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { Card, TextArea } from 'semantic-ui-react';
import { setTab } from '../actions';
import '../styles/feedback.css';
import icon from '../assets/icon/comments-solid.svg';
import HttpService from './../services/HttpService';
import { Redirect } from 'react-router-dom';
import { Dialog } from '@material-ui/core';
/** URL for API */
const URL = process.env.REACT_APP_API;

/**
 * feedback page
 */
class Feedback extends PureComponent {
  /**
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);
    /**
     * feed back service
     */
    this.service = new HttpService();
    /**
     * state
     * show: boolean to show pop up or not
     * feedback: user feedback
     */
    this.state = {
      show: false,
      feedback: '',
      err: false,
      goHome: false,
    };
  }

  componentDidMount() {
    this.props.dispatch(setTab({ tab: 'null' }));
  }
  /**
   * send the feedback to feedback service
   */
  sendFeedback = async () => {
    const { feedback } = this.state;
    if (feedback.trim() === '') {
      this.setState({ err: true });
      return null;
    }
    const url = `http://${URL}/feedback`;
    const body = { feedback: feedback };
    const result = await this.service.postRequest(url, {}, {}, body);
    if (result && result.key) {
      this.setState({ show: true });
    } else {
      this.setState({ sendErr: true });
    }
  };

  /**
   * handle feedback content changes
   */
  handleFeedbackChange = (event) => {
    this.setState({ feedback: event.target.value });
  };

  home = () => {
    this.setState({ goHome: true });
  };

  /**
   * renders everything for feedback page
   */
  render() {
    const { show, err, goHome, sendErr } = this.state;
    if (goHome) {
      return <Redirect to="/"></Redirect>;
    }

    return (
      <div>
        <Card className="feedback-card">
          <div className="feed-header">
            <img src={icon} alt="comments" id="feedback-img" />
            <div className="feedback-title">Website Feedback</div>
            <div className="sub-text">
              Please give any relevant feedback about the website and user
              experience, or provide a suggestion on what could be better. This
              feedback will be forwarded to our development team directly.
            </div>
          </div>

          <TextArea
            placeholder="Help us improve!"
            className="feedback-text"
            style={{ minHeight: 100, maxWidth: '100%', minWidth: '100%' }}
            onChange={this.handleFeedbackChange}
          />
          <div>
            {err ? (
              <div className="ui negative message feed-err">
                Please fill out the comment box.
              </div>
            ) : (
              ''
            )}
            <div className="ui button feedback-btn" onClick={this.home}>
              Home
            </div>

            <div
              className="ui blue button feedback-btn"
              onClick={this.sendFeedback}
            >
              Submit
            </div>
          </div>
        </Card>
        <Dialog className="feed-dialog" open={show}>
          <div className="dialog-header">Thanks for your feedback!</div>
          <div className="ui blue button" id="home" onClick={this.home}>
            Home
          </div>
        </Dialog>
        <Dialog className="feed-dialog" open={sendErr}>
          <div className="dialog-header">
            Oops! Something went wrong sending your feedback. Comeback later and
            try again.
          </div>
          <div className="ui blue button" id="home" onClick={this.home}>
            Home
          </div>
        </Dialog>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  // view: state.view,
});

export default connect(mapStateToProps)(Feedback);
