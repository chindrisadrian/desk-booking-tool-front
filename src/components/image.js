import React, { Component } from 'react';
import '../styles/base.css';

class Image extends Component {
  constructor(props) {
    super(props);
    this.state = {
      img: '',
    };
  }

  arrayBufferToBase64(buffer) {
    let binary = '';
    const bytes = [].slice.call(new Uint8Array(buffer));
    bytes.forEach((b) => (binary += String.fromCharCode(b)));
    return window.btoa(binary);
  }

  componentDidMount() {
    const base64Flag = 'data:image/jpeg;base64,';
    const imageStr = this.arrayBufferToBase64(this.props.data); // data.img.data.data);
    this.setState({
      img: base64Flag + imageStr,
    });
  }

  render() {
    const { img } = this.state;
    return <img className="floor-pic" src={img} alt="Loading" />;
  }
}
export default Image;
