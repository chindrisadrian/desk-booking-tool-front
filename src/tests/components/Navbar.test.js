import React from 'react';
import { Navbar } from '../../components/NavBar';
import { shallow } from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils'; // ES6

//jest.mock('../../services/Auth');

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Navbar />);
});

test('Should render when logged in', () => {
  //wrapper.setProps({ loggedIn: true });
  expect(wrapper).toMatchSnapshot();
});
