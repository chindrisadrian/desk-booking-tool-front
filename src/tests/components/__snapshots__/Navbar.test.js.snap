// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render when logged in 1`] = `
<div>
  <div
    className="site-header"
  >
    Desk Booking Tool
  </div>
  <div
    className="site-subheader"
  >
    User Panel
  </div>
  <div
    className="custom-divider"
  >
     
  </div>
  <div
    className="ui blue six item menu"
  >
    <Link
      className="item"
      replace={false}
      to="/BookADesk"
    >
      Book A Desk
    </Link>
    <Link
      className="item"
      replace={false}
      to="/ViewPreviousBookings"
    >
      View My Previous Bookings
    </Link>
    <Link
      className="item"
      replace={false}
      to="/ViewAllBookings"
    >
      View All Bookings
    </Link>
    <Link
      className="item"
      replace={false}
      to="/userRestrictions"
    >
      Timing Restrictions
    </Link>
    <Link
      className="item"
      replace={false}
      to="/restrict"
    >
      Restrict Desks
    </Link>
    <Link
      className="item"
      replace={false}
      to="/addFloor"
    >
      Add a Floor
    </Link>
  </div>
  <div
    className="custom-divider"
  >
     
  </div>
</div>
`;
